DOC: Change selected hardlinks to NEPs to intersphinx mappings

eagunn opened the pull request:
This commit changes links to NEPs from explicit https urls to intersphinx :ref:s. Selected files are 'actual doc files' (my phrase), not release notes, change logs, or the NEPs themselves.

These changes are a first installment towards addressing #26707. 

I plan on making the same sort of changes to the NEP files themselves, where most of the hardlinks are found. But I'd like to get this first PR through review and acceptance first as I'm new to the development workflow.

[skip actions] [skip azp] [skip cirrus]

eagunn committed:
DOC: Change selected hardlinks to NEPs to intersphinx mappings

This commit changes links to NEPs from explicit https urls to intersphinx :ref:s.
Selected files are 'actual doc files' (my phrase), not release notes,
change logs, or the NEPs themselves.

These change are a first installment towards addressing #26707.

[skip actions] [skip azp] [skip cirrus]

InessaPawson reviewed (approved):
Looks ready to me. Thank you for working on it, @eagunn!

melissawm commented on doc/source/dev/development_environment.rst:
```diff
@@ -327,7 +327,7 @@ typically packaged as ``python-dbg``) is highly recommended.
 .. _`match test names using python operators`: https://docs.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests
 .. _`Python Style Guide`: https://www.python.org/dev/peps/pep-0008/
-.. _`C Style Guide`: https://numpy.org/neps/nep-0045-c_style_guide.html
+.. :ref:`NEP45`
```
The reference is already selected above, so you don't need this line:
```suggestion
```

eagunn replied:
Good catch, @melissawm. I didn't understand what those reference definition lines were doing at the bottom of the file and should have asked or searched for an explanation before I checked in. That line has been removed.

melissawm reviewed (commented):
One minor comment - thanks @eagunn !

eagunn committed:
DOC: Remove redundant reference definition

As pointed out by @melissawm, the NEP45 reference is predefined and
does not need to have a line in the reference definition section at
the bottom of the document.

This correction is the next installment towards addressing #26707.

[skip actions] [skip azp] [skip cirrus]

melissawm commented:
Something is wrong with CI but it's not you - see https://github.com/numpy/numpy/issues/26820

mattip commented:
Rebased off main to get the fix for #26820.

mattip commented:
Thanks @eagunn
